### PR TITLE
Issue #1076 - support wildcard globs for project sources & destinations

### LINKED ISSUE #1076: Support wildcard globs for project sources & destinations
The IsSourcePermitted and IsDestinationPermitted functions only support either a `*` wildcard, or an explicit source/destination. It does not support wildcard globs (e.g. `apps-*`).

### PR DESCRIPTION
**Pull request by alexmt**
Closes #1076

### CODE REVIEW

File: pkg/apis/application/v1alpha1/types_test.go [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+	}, {
+		projSources: []string{"ssh://git@GITHUB.com:argoproj/test"}, appSource: "ssh://git@github.com:argoproj/test", isPermitted: true,
+	}, {
+		projSources: []string{"https://github.com/argoproj/*"}, appSource: "https://github.com/argoproj/argoproj.git", isPermitted: true,
+	}, {
+		projSources: []string{"https://github.com/test1/test.git", "https://github.com/test2/test.git"}, appSource: "https://github.com/test2/test.git", isPermitted: true,
+	}, {
+		projSources: []string{"https://github.com/argoproj/test1.git"}, appSource: "https://github.com/argoproj/test2.git", isPermitted: false,
+	}, {
+		projSources: []string{"https://github.com/argoproj/*.git"}, appSource: "https://github.com/argoproj1/test2.git", isPermitted: false,
```
Comment (jessesuen) [inline pkg/apis/application/v1alpha1/types_test.go]: Can you verify that the following is *not* permitted?
```
projSources: []string{"https://github.com/argoproj/foo"}, appSource: "https://github.com/argoproj1/foo1", isPermitted: false,
```

File: pkg/apis/application/v1alpha1/types_test.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+			Server: "https://*.default.svc", Namespace: "default",
+		}},
+		appDest:     ApplicationDestination{Server: "https://kubernetes.default.svc", Namespace: "default"},
+		isPermitted: true,
+	}, {
+		projDest: []ApplicationDestination{{
+			Server: "https://team1-*", Namespace: "default",
+		}},
+		appDest:     ApplicationDestination{Server: "https://test2-dev-cluster", Namespace: "default"},
+		isPermitted: false,
```
Comment (jessesuen) [inline pkg/apis/application/v1alpha1/types_test.go]: Can we add tests to verify namespace globbing is working?

```
		projDest: []ApplicationDestination{{
			Server: "https://kubernetes.default.svc", Namespace: "test-*",
		}},
		appDest:     ApplicationDestination{Server: "https://kubernetes.default.svc", Namespace: "test-foo"},
		isPermitted: true,
```


```
		projDest: []ApplicationDestination{{
			Server: "https://kubernetes.default.svc", Namespace: "test-*",
		}},
		appDest:     ApplicationDestination{Server: "https://kubernetes.default.svc", Namespace: "test"},
		isPermitted: false,
```

File: pkg/apis/application/v1alpha1/types_test.go (hunk 1 above)
Comment (alexmt) [reply pkg/apis/application/v1alpha1/types_test.go]: Done

File: pkg/apis/application/v1alpha1/types_test.go (hunk 2 above)
Comment (alexmt) [reply pkg/apis/application/v1alpha1/types_test.go]: Done

**Review by jessesuen — APPROVED** [review]

**Pull request merged**